packages/sync/src/webrtc-http-stream-signaling.js

```diff
@@ -0,0 +1,401 @@
+									}
+								}
+								if ( data.to === peerId ) {
+									map.setIfUndefined(
+										webrtcConns,
+										data.from,
+										() =>
+											new WebrtcConn(
+												signalCon,
+												false,
+												data.from,
+												room
+											)
+									).peer.signal( data.signal );
+									emitPeerChange();
+								}
+								break;
+						}
+					};
+					if ( room.key ) {
+						if ( typeof m.data === 'string' ) {
+							cryptoutils
+								.decryptJson(
+									buffer.fromBase64( m.data ),
+									room.key
+								)
+								.then( execMessage );
+						}
+					} else {
+						execMessage( m.data );
+					}
+				}
+			}
+		}
+	);
+	signalCon.on( 'disconnect', () => log( `disconnect (${ url })` ) );
+}
+
+/**
+ * Method that instantiates the http signaling connection.
+ * Tries to implement the same methods a websocket provides using ajax requests
+ * to send messages and EventSource to retrieve messages.
+ *
+ * @param {HttpSignalingConn} httpClient The signaling connection.
+ */
+function setupHttpSignal( httpClient ) {
+	if ( httpClient.shouldConnect && httpClient.ws === null ) {
+		// eslint-disable-next-line no-restricted-syntax
+		const subscriberId = Math.floor( 100000 + Math.random() * 900000 );
+		const url = httpClient.url;
+		const eventSource = new window.EventSource(
+			addQueryArgs( url, {
+				subscriber_id: subscriberId,
+				action: 'gutenberg_signaling_server',
+			} )
+		);
+		/**
+		 * @type {any}
+		 */
+		let pingTimeout = null;
+		eventSource.onmessage = ( event ) => {
+			httpClient.lastMessageReceived = Date.now();
+			const data = event.data;
+			if ( data ) {
+				const messages = JSON.parse( data );
+				if ( Array.isArray( messages ) ) {
+					messages.forEach( onSingleMessage );
+				}
+			}
+		};
+		// @ts-ignore
+		httpClient.ws = eventSource;
+		httpClient.connecting = true;
+		httpClient.connected = false;
+		const onSingleMessage = ( /** @type {any} */ message ) => {
+			if ( message && message.type === 'pong' ) {
+				clearTimeout( pingTimeout );
+				pingTimeout = setTimeout(
+					sendPing,
+					messageReconnectTimeout / 2
+				);
+			}
+			httpClient.emit( 'message', [ message, httpClient ] );
+		};
+
+		/**
+		 * @param {any} error
+		 */
+		const onclose = ( error ) => {
+			if ( httpClient.ws !== null ) {
+				httpClient.ws.close();
+				httpClient.ws = null;
+				httpClient.connecting = false;
+				if ( httpClient.connected ) {
+					httpClient.connected = false;
+					httpClient.emit( 'disconnect', [
+						{ type: 'disconnect', error },
+						httpClient,
+					] );
+				} else {
+					httpClient.unsuccessfulReconnects++;
+				}
+			}
+			clearTimeout( pingTimeout );
+		};
+		const sendPing = () => {
+			if (
+				httpClient.ws &&
+				httpClient.ws.readyState === window.EventSource.OPEN
+			) {
+				httpClient.send( {
+					type: 'ping',
+				} );
+			}
+		};
+		if ( httpClient.ws ) {
+			httpClient.ws.onclose = () => {
```

Comment:
Being the `httpClient.ws` an [EventSource](https://developer.mozilla.org/en-US/docs/Web/API/EventSource), it doesn't seem that it has an `onClose` event. Do we still need this code?

Reply:
I think we may still need this code httpClient.ws does not have an onClose event and this code is adding an onclose function to it in case y-wertc calls the onclose method (y-webrtc expects httpClient.ws to be a websocket which has an onclose event).